spark/src/main/scala/com/gu/tableversions/spark/VersionedFileSystem.scala

```diff
@@ -0,0 +1,159 @@
+package com.gu.tableversions.spark
+
+import java.net.URI
+import java.util.Objects
+
+import com.gu.tableversions.core.Partition
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+abstract class ProxyFileSystem extends FileSystem {
+
+  protected var pathMapper: PathMapper = _
+
+  def setPathMapper(mapper: PathMapper): Unit =
+    pathMapper = mapper
+}
+
+class VersionedFileSystem extends ProxyFileSystem {
+
+  private var baseURI: URI = _
+  private var baseFS: FileSystem = _
+  override def initialize(name: URI, conf: Configuration): Unit = {
+    val cacheDisabled = conf.getBoolean("fs.versioned.impl.disable.cache", false)
+    val baseFSScheme = conf.get("fs.versioned.baseFS")
+    val version = conf.get("fs.versioned.version")
+
+    require(Objects.nonNull(baseFSScheme), "fs.versioned.baseFS not set in configuration")
+    require(Objects.nonNull(version), "fs.versioned.version not set in configuration")
+    require(cacheDisabled, "fs.versioned.impl.disable.cache not set to true in configuration")
+
+    import scala.collection.JavaConverters._
+
+    val partitionMappings = conf.iterator.asScala
+      .filter(_.getKey.startsWith(VersionedFileSystem.keyPrefix))
+      .map(_.getValue)
+      .toList
+      .map { v =>
+        (v, v + "/" + version)
+      }
+      .toMap
```

Comment:
It should be good to factor out this code into a separate method, then unit testing that. (Or maybe not the bit that interacts with the Hadoop config, but everything else?)